### ISSUE TITLE
Thoughts on automated moderation tooling

### ISSUE THREAD

**Issue by PawanHegde** [LemmyNet/lemmy#3281]
# Context
I’m interested in working on improving the experience for moderators. I’m especially looking at auto-moderation systems. Think of reddit's `automoderator` or the [vast number of bots](https://www.reddit.com/r/modguide/wiki/bots/) people have written for reddit. 

I wanted to start a separate project with moderation bots that lemmy mods could host and run on their own, but the more I think about it, the more I feel like we need a hybrid ecosystem with at least some auto-moderation tools bundled with lemmy. I want to get to some kind of consensus on which moderation tools are important for lemmy and which of those should come bundled with lemmy itself.

>:warning: I might update this top-level post based on the discussion in the issue below.

## Goals
- Communities should be able to scale as more people join lemmy communities without moderation being a bottleneck.
- Moderators should be able to customize auto-moderation tools without technical know-how.
- Developers know what tools to build so that there is full coverage and not too much duplication for the same few capabilities.

## Intended audience
- **The current active contributors to lemmy**: I'm new to rust, lemmy, **and** ActivityPub so I'd really appreciate guidance.
- **Current mods on Lemmy or people who mod/used to mod large subreddits**: What functionality are you missing in lemmy?

### Lemmy
There is no official tool for moderating; enterprising people have [increasingly](https://github.com/ornato-t/lemmy-automoderator) started [writing](https://github.com/skyraptor777/Lemmy_bots) their [own](https://github.com/thomasdouwes/lemmy-reply-bot) moderation [bots](https://github.com/ShittyKopper/lemmod). 

### Reddit
The [in-built Automoderator](https://www.reddit.com/wiki/automoderator/full-documentation/#wiki_actions) is by far the most commonly used moderation tool on reddit. There are also [lots of other bots](https://www.reddit.com/r/modguide/wiki/bots/), but they mostly have limited scope and are run by volunteers.

# Proposal
## `Automoderator` for Lemmy (`Janitor`?)
We add an analogue of the automoderator in core lemmy repo. This shouldn't (and can't) be a one-to-one mapping of the automoderator—Lemmy has different challenges—but we can look at automoderator and some of the other popular reddit bots for inspiration. See the table below for a high-level idea of what functionalities they have.

Janitor should only be responsible for moderation functionalities that
1. Only need context already present in the current server. I want to discuss what this exactly means in the fediverse because users can post/comment on communities in other servers. But for example, it should be totally out of scope for the janitor of an server to call other servers. We might be able to use external bots (next section) for that.
3. Can be performed by regular mods. At no point should the janitor have more access than regular moderators.

To use Janitor, the only action needed by mods to use it should be to provide configuration. They would **not** need to also add it to the moderators' list. We will need to special-case Janitor here. 

Other than the special-casing above, **Janitor should only talk to the backend via the official API** so that we don't need to test another code path and it won't have access that regular mods don't. I'd actually prefer it it ran as a separate process like pict-rs so that if in the future if some admins want to run it on separate machine for scaling purposes, we _can_ support that.

It'll also be easier to extend Janitor with new functionality if we see good open-source modbots.

## Other modbots (non-Janitor)
Any check that is experimental, _non-essential_, or would take up an inordinate amount of resources on the main server would need to be hosted voluntarily by mods using their own resources. In the case of reddit, I see owners of the modbots running their own machines and choose which communities can sign up. We can follow a similar model.

Admins/Mods can host modbots on their own or share it with other mods if they wish to. To use a modbot, user will need to create an account for the bot if there isn't one already for the server that the community is on and add it as a moderator. This is to handle deferation. But the same modbot instance can be used for multiple servers.

While we don't want Janitor to talk to other servers, independent modbots can listen in on multiple servers to form their own read model. Independent modbots can detect bad actors that have their own servers and spam multiple communities on different servers in parallel.

## Configuration
This is crucial. We should introduce a standard location for configuration. This will definitely apply to `Janitor` but we should encourage other modbots to also use the same scheme.

For example, if you moderate a community called `foodom` and you want to configure `Janitor`, you'll post a config (in YAML) to `/c/foodom/bot/janitor`. (Where possible, we should try to support the same [config options](https://www.reddit.com/wiki/automoderator/writing-basic-rules/) as `automoderator` so that reddit communities can copy over their configs and expect it to work roughly the same way.)

Other modbots can add their own locations like `/c/foodom/bot/spamocop`. But it's important that we support this endpoint in lemmy so that we can maintain authorization. Only the mods can access the configs.

If we don't have a place to host configuration in a way that can be controlled by the mods, modbots would need to add the configs to some other location. Every bot-writer will need to think about hosting their configs and we'll soon have multiple ways to do it. Some might choose to use environmental variables, some might use external repos, and some other hacks possibly. This makes it hard for mods to onboard overall. Also, the configs will be owned by whoever is running the bot. So adding/removing mods to a community becomes a two-step process. You'd need to add/remove them on lemmy and then add/remove their access to the configs for the bot. Without that, ex-mods will have elevated access to communities.

### Why not add all moderation directly to lemmy and not allow any modbots?
We cannot unilaterally decide that there will be no other bots. There are some things that are out of scope for us. There are also some moderation techniques that are useful for only some communities. We cannot centrally decide for all.

### Why special-case Janitor? Using modbots seems like a general enough solution.
That is true conceptually and would totally work in an ideal world, but it increases friction enough to turn off moderators.
- Mods would need to think about and pay for hosting the bots. And also coordinate accounts for these bots.
- Adds a new attack vector because the bot set up by the mods can also be compromised. We have more control over code in the lemmy codebase.
- More than one bot can be added by different mods to the same community. Potential for hilarity due to race conditions. This can happen with Janitor and other bots, but if Janitor is good enough, mods wouldn't want to add a new modbot.

# Appendix
| Functionality| Needed in Lemmy | Supported in Reddit | Comment|
|-------------|--------------------|---------------------|----------|
| Approve or Reject post/comment | Yes | Automoderator | Some conditions for triggering action are not relevant, but let's not get into that detail now |
| Set/overwrite flair | No | Automoderator | Not needed until lemmy [starts supporting post tags](https://github.com/LemmyNet/lemmy/issues/317), but this will be a really useful feature to have |
| Set sticky | Yes | Automoderator | Lemmy calls this pinning |
| Set locked | Yes | Automoderator | |

> :warning: Table actively being updated

**Comment by PawanHegde** [comment]
@SleeplessOne1917 / @dessalines / @Nutomic: I made the mistake of creating this as a blank issue. Could one of you please add the appropriate label to this? `discussion`/`proposal`?

**Comment by boehs** [comment]
<img width="783" alt="image" src="https://github.com/LemmyNet/lemmy/assets/51836263/de70aab8-f278-4eec-9b2d-6b5d6955ab9f">

Also see #3277, which I opened a couple of hours ago (funny how that goes)

Key differences:

1. This proposal doesn't seem to define the configuration language? Will it mirror the pesudo-yaml of reddit? #3277 proposes a lightweight programming language
2. This proposal suggests hosting the moderation bot externally (so `post -> post processed on backend -> API sent to bot -> bot responds over API`), #3277 proposes embedding these capabilities directly into the source code (`post -> post processed on backend -> backend step where modbot code code runs` (the rational for this is described in #3277)

Otherwise, implementing one closes both. Worth discussing the pros and cons, intended paths forward, and if they can be merged

**Comment by boehs** [comment]
For what it's worth

1. I believe embedding this capacity into the main server is better than an independent bot (some rational is in #3277)
2. Configuration language is more nuanced.

When I was a moderator on Reddit, automod was hard to learn, hard to work with, and sometimes not flexible enough (or, at the very least, resulted in clunky gigantic regexes). I'm not particularly keen on keeping stuff the same for compatibility, though I understand people's thoughts differ. A programming language would further alleviate the need for 3rd party bots. Additionally, I presume the backend implementation will be more complex for an automod YAML thing.

That being said, I understand the perspective of wanting it to be easy for reddit moderators to port their configurations over. Also, introducing a language does incur a risk. I'm pretty confident in rhai's ability to maintain the sandbox, but not as much in it's ability to limit execution time and RAM consumption (though there are some tools to deal with this 95% or so). I've detailed all my thoughts on this in #3277. YAML does not face this same issue (in theory, not really, see https://en.wikipedia.org/wiki/ReDoS)

**Comment by 0xAnansi** [comment]
Due to the architecture of lemmy, I would suggest that the external hosting of the bot should be possible, even speaking for a bot only managed by the administrators of the instance.
The bot actions will be API/IO intensive, scaling with the number of communities and rules, which could tank the instance if it's hosted on the same server.

**Comment by PawanHegde** [comment]
Thanks for the comments, @boehs! I just went through #3277. Great to see other people also interested in this topic. 🙂   

#3277 proposes that mods will write `Rhai` scripts with DSL functions like `ban()`, `pin()`,  `prevent()` etc. The custom functions would be exposed by core lemmy and the script would be run synchronously before each post or comment is posted. Some issues I see are:
- [Rhai is about 2x slower than Python 3](https://rhai.rs/book/about/benchmarks.html) and I'd expect it to be magnitudes slower than something like normal Rust. This is going to matter when our instances get overloaded and we would need to run custom scripts written by mods before posting comments. Note: `Rhai` has a [syntax like `Rust` and `Javascript`](https://rhai.rs/book/about/features.html) and is run dynamically.
- The embedded scripts will need to be called synchronously if we want to allow `prevent()`. This will be a bottleneck, especially during a surge. If we were to run the script asynchronously in a queue, then the bot could as well have been an external process. There is no need to embed the engine in core lemmy.
- Mods need to learn [Rhai](https://rhai.rs/book/about/non-design.html). This is going to need a much steeper learning curve than YAML. I'm afraid that some mods will give up because of the syntax or write terrible scripts. If we use YAML syntax like automoderator, existing reddit mods will be able to carry over their knowledge even if the configs aren't exactly the same.
- The checks in automod can be implemented once and used by everyone; we can keep improving them over time and all users will benefit. But if mods are forced to write `Rust` (`Rhai`) code for moderation, I am afraid that we'll have a hundred different buggy ways of doing the same thing and no way for us to fix them en-masse. We will also be able to protect lemmy instances from incompetence and/or malice of bad scripts.
- Embedded `Rhai` is so different from bots that there is no interoperability and no way to move their functionality to/from core. With the approach in this proposal, if there are open-source bots that are very popular with lemmy communities, we can choose to run them along with the instance and all mods get this functionality out of the box with no change in their configs. With the approach in #3277, I guess the equivalent would be implementing a Rust function and asking mods to update their scripts?
- An embedded script _can_ have elevated privileges because it will be interacting with core lemmy directly instead of via the API. Even `prevent` is an action that regular moderators can't do so we are starting with elevant privileges off the bat.
- Because the script will have a different point of entry and code-path compared to regular moderators, it's easier for them to diverge. With an API, you know that both are using the same mechanism.

That said, `Rhai` will be more flexible for powermods than the approach in this proposal. I'd like to think that we're trading off flexibility for safety and scalability.

>implementing one closes both

The scope for this proposal is slightly larger than #3277, so no matter what route we take for the in-built Automoderator, I'd like us to remember that I also want to discuss (a) Configuring non-built-in bots (b) Which moderation functionalities are needed.

**Comment by PawanHegde** [comment]
> @0xAnansi The bot actions will be API/IO intensive, scaling with the number of communities and rules, which could tank the instance

True. Your comment reminded me that we also need a way to queue up moderation requests so that we can rate-limit and smoothen the curve. I'll add it to the main proposal. Thanks.

**Comment by ornato-t** [comment]
An in-site automatic moderator tool would surely be great! I [created](https://github.com/ornato-t/lemmy-automoderator) an automod repo because I am not familiar enough with the lemmy codebase and I expected a huge influx of Reddit users and the need for mod tools to help moderators keep everything under control, so I planned on creating a "quick and dirty" solution to get rid of the most heinous of spam in the first phases of the Reddit migration. What you are suggesting would certainly be better in the long run.

I haven't started working on the bot yet, as my pipeline is very full at the moment. I think I'll get to it in the next weeks. The idea behind my automod was an open source, self hosted bot that every instance admin could spin up on its own, though I see how this might be a bottleneck for larger instances with millions of users.

I think the most feasible solution would be relying on smaller community built bots as a first phase and then move on to something more stable, embedded in the lemmy server itself and configurable via the UI in the future. For what is worth, I'll try following the specs you listed as closely as possible when building my automod.

**Comment by SleeplessOne1917** [comment]
> I think the most feasible solution would be relying on smaller community built bots as a first phase

I made a library (which I still need to update for 0.18) just for that sort of thing.

**Comment by PawanHegde** [comment]
Hey @ornato-t and @SleeplessOne1917! 

> enterprising people have [increasingly](https://github.com/ornato-t/lemmy-automoderator) started [writing](https://github.com/skyraptor777/Lemmy_bots) their [own](https://github.com/thomasdouwes/lemmy-reply-bot) moderation [bots](https://github.com/ShittyKopper/lemmod).

Small 'verse. I'd mentioned both your projects in the initial proposal 😄 Also paging @ShittyKopper as another person working on their own bot to get their opinion.

>I think the most feasible solution would be relying on smaller community built bots as a first phase and then move on to something more stable, embedded in the lemmy server itself and configurable via the UI in the future

IMHO it's good to have an ecosystem of independently-run community-owned bots even after some fundamental bots are bundled with core-lemmy. If some of these independent bots become eally popular, we can (if the project owner agrees) bundle them with lemmy. So we don't need to think of this as phase-1 and phase-2 but as complementary. Also, I'd rather see the bots "bundled" with lemmy and not "embedded" into lemmy itself so that they can be hosted separately by the admins if needed (hosting being transparent to the mods).

I think of "automoderator for lemmy/Janitor" as standard bots that come bundled with lemmy so that all they need are configurations.

I think the only thing we need to standardise quickly is configurations (host where? what format?) and maybe a list of capabilities that we need so that interested devs _can_ start working on them parallely.

Reddit uses the format `/r/foo/wiki/config/<bot-name>`. Everything under `config` is restricted to just mods while the other wiki pages are not. This seems tacked on. We can instead have a separate route for bot configs.

What do you all think about YAML as the format for configurations and `PUT /api/v3/community/<community-name>/bot/<bot-name>` on the backend & `/c/<community-name>/bot/<bot-name>` as the location for storing the bot configurations? I can get started working on a PR to add the backend route. I'll need help with the frontend. The routes should be restricted to only mods.

**Comment by 0xAnansi** [comment]
@PawanHegde the storage of settings under /r/subname/wiki/config/botname is a convention and not an obligation for all bots but automod afaik.

I don't think we should consider the lemmy automod differently than any other bot. It should, imo, be a separate project that can be installed on the same instance or outside the instance to avoid too big of a load at the same place, especially for big instances.

If we go for that, it means that we first need to come out with a data storage equivalent to reddit's wiki for each community.

For the format, the easiest way would be to be able to use an automod config for the bot, so moderators can seemlessly migrate their automod config from reddit to lemmy.

**Comment by ShittyKopper** [comment]
Some of the issues I encountered building mine (which now that I have selfhosting set up I may bring up a public instance of) is that:

- It is impossible to track edits without re-checking each post and comment manually
- Overall clarification on how the API works (aka "better docs in general")
    - Clarification on reliability of the built-in mark as read functionality. Will something as simple as fetching a new post mark it as read?
    - Clarification on how exactly IDs work. Especially on federated resources.
    - Clarification on how "trusted" certain parts of the API are. In particular, actor_id seems to be the only way to get a "textual" representation on the instance something came from. Will that always be a URL or can a malicious remote server mess with it?
    - Clarification on permissions. Right now, the "rename" action of my bot will not work, as moderators do not have the ability to rename things. This isn't written anywhere in the documentation.
- Either the current pagination system needs to change, or an additional one needs to be added
    - Both Reddit and Tildes (whose dev worked on Reddit and AutoModerator before it) use an "x posts/comments after post id" style of pagination, so posts don't jump around as newer ones get added
- Do something about rate limits in general, even if it's just vague approximations on how resource intensive certain API calls are.

**Comment by brettwilcox** [comment]
I wanted to provide some thoughts from 13 years of moderating /r/usenet.

As @PawanHegde [mentioned](https://github.com/LemmyNet/lemmy/issues/3281#issuecomment-1607267882), there is a wiki page that is ultimately parsed and used in the automod config.

## /r/usenet automoderator config (2023)

Here is an export of the /r/usenet automoderator config as it stands today. - https://github.com/scorpion/usenet/blob/main/automoderator/USENET.md

Here is an image of the log with history - https://github.com/scorpion/usenet/blob/main/automoderator/usenet-mod-log.png

## Reddit automoderator script (2017)

Here is reddit's automoderator.py script that used to parse these scripts - https://github.com/reddit-archive/reddit/blob/753b17407e9a9dca09558526805922de24133d53/r2/r2/lib/automoderator.py

There have likely been lots of developments around that script since it went closed source in ~2017.

I'm going to have a heck of a time recruiting folks to moderate if there is not a managed automated solution in place.

**Comment by brettwilcox** [comment]
Some further thoughts. I think under the hood, both github and reddit use [gollum](https://github.com/gollum/gollum). What if there was another recommended package for wiki software based on rust like [sausagewiki](https://github.com/maghoff/sausagewiki).

We could find a way to implement and import the rules based on communities. It needs integration at the UI level for mods to access. It also needs the ability to log and roll back bad moderation rule changes.

**Comment by ornato-t** [comment]
@brettwilcox as others have suggested, I don't think we need to limit ourselves to the wiki structure that Reddit used, especially because we would be implementing it from scratch. We can make our own API endpoint instead, maybe even a mod only one (why let average user see the AutoMod config if they don't need to?).

I do however appreciate your insights as an experienced mod, as a user I feel the need for mod tools but hardly have any idea of how those tools should look like.

@PawanHegde the specifics of the API call aren't terribly important right now, but I agree on the point that we should recycle YAML for the configuration. I think making this accessible for escaping Reddit moderators should be a priority. We probably won't be able to make it compatiblle with the Reddit AutoMod but we can at least make it look as similar as possible in their eyes.

**Comment by snowe2010** [comment]
Should the focus not be on creating the api we want, and then create a translation layer for standard automod config to translate those concepts to lemmy's api concepts? I do not think we should limit ourselves to yaml.

**Comment by PawanHegde** [comment]
Thank you everyone.

> @ShittyKopper: It is impossible to track edits without re-checking each post and comment manually

This is relevant for bots. Users might edit their comments after the automoderator has already performed an action on them. 

The other points you raised are all excellent and deserve separate issues (if there aren't any already); they would benefit general development on lemmy. But I'm not qualified to address these, so I'm tagging @dessalines and @Nutomic to get their attention.

**Comment by PawanHegde** [comment]
> @snowe2010: I do not think we should limit ourselves to yaml.

I think I'd have gone with YAML even if there was no reddit automoderator because of how ubiquitous YAML is for configs. I could be wrong, but I don't think we need to translate the configs because mods on lemmy would need similar options as on reddit.

This is an example from https://www.reddit.com/wiki/automoderator/writing-basic-rules/
```
domain: [youtube.com, youtu.be]
comment: |
    Sorry, your submission has been automatically removed.

    This subreddit does not allow videos to be posted.

    You might want to consider posting to /r/videos instead.
action: remove
```
They only declare the intent. The bot can implement it as it pleases.

**Comment by PawanHegde** [comment]
> @brettwilcox:  It also needs the ability to log and roll back bad moderation rule changes.

That's a great point. I had been thinking of only storing the latest configs, but this should be a better UX.

**Comment by BarterClub** [comment]
As a mod we really need this before we get to large. There is a large amount of spam when I was on reddit. 

The big feature is to prevent spam from brand new accounts and bots using the same words when they post.

**Comment by Nutomic** [comment]
[LemmyAutomod](https://github.com/RikudouSage/LemmyAutomod) exists now. If there are moderation features missing from Lemmy please open specific issues. This one is too vague to act on.

**Comment by SteveWBausch** [comment]
I first admit that a great idea, "Bell the cat", begs the question, "who's going to bell the cat?"

The site, and my community, needs a modest revenue stream to maintain itself. Why not a billing mechanism for inane post titles, obscenities and links?

Let the childish amongst us pay for the site.

And a mechanism to delete such posts after 12 hours or so.

For repeat offenders, I suggest an exponential fee increase for subsequent postings.

Can't be worse than the silly boilerplate generated by automated moderation.

**Comment by dessalines** [comment]
That has nothing to do with this issue, and we're not going to add any monetization directly into lemmy.
